src/Lean/Meta/KAbstract.lean

```diff
@@ -41,15 +43,22 @@ def kabstract (e : Expr) (p : Expr) (occs : Occurrences := .all) : MetaM Expr :=
         visitChildren ()
       else if e.toHeadIndex != pHeadIdx || e.headNumArgs != pNumArgs then
         visitChildren ()
-      else if (← isDefEq e p) then
-        let i ← get
-        set (i+1)
-        if occs.contains i then
-          return mkBVar offset
+      else
+        -- We save the metavariable context here,
+        -- so that it can be rolled back unless `occs.contains i`.
+        let mctx ← getMCtx
+        if (← isDefEq e p) then
```

Comment:
What about exceptions? We should handle them here, or add a comment saying why it is ok to **not** rollback the mctx in this case.

Reply:
Sorry, I don't understand how exception handling is relevant here.

* Previously, if there was an exception at `isDefEq` then that would propagate to the caller of `kabstract`. That is still the case. Why would we try handling the exception here now?

* We are now rolling back the mctx **if** the occurrence has been rejected by the caller. Why would we want to rollback the mctx on an exception? (Again, previously an exception in `isDefEq` would result in the caller of `kabstract` receiving whatever mctx resulted.)

Reply:
@leodemoura, could I ping you on this query?